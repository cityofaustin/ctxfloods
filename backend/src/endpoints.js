
var endpoint = 'http://localhost:5000/graphql';

if(process.env.USE_AWS_ENDPOINT) {
  endpoint = process.env.POSTGRAPHQL_ENDPOINT;
};

module.exports = {
    endpoint: endpoint
};
