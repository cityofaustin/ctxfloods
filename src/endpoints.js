
var endpoint = 'http://localhost:5000/graphql';

if(process.env.CI) {
  endpoint = process.env.AWS_ENDPOINT;
};

module.exports = {
    endpoint: endpoint
};
