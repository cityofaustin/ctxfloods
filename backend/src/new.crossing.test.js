import HttpTransport from 'lokka-transport-http';
import Lokka from 'lokka';
import { endpoint } from './endpoints';

const anonLokka = new Lokka({transport: new HttpTransport(endpoint)});
const superAdminEmail = 'superadmin@flo.ods';
const communityAdminEmail = 'admin@community.floods';
const communityEditorEmail = 'editor@community.floods';
const everyPassword = 'texasfloods';
const longitude = -97.755996;
const latitude = 30.30718;
const newCrossingMutation = `
  mutation($communityId:Int!, $longitude:Float!, $latitude:Float!) {
    newCrossing(input: {
      name: "New Crossing"
      humanAddress: "In test land"
      description: "TEST LAND IS MAGIC!"
      communityId: $communityId
      longitude: $longitude
      latitude: $latitude
    }) {
      crossing {
        id
      }
    }
  }
`;

async function getToken(email, password) {
  const response = await anonLokka.send(`
    mutation($email:String!, $password:String!) {
      authenticate(input: {email: $email, password: $password}) {
        jwtToken
      }
    }
  `,
  {
    email: email,
    password: password
  });

  return response.authenticate.jwtToken;
}

function shouldWork(email, password, communityId, longitude, latitude, extra_description) {
  describe('as ' + email + ' ' + (extra_description || ''), () => {
    var lokka;

    beforeAll(async (done) => {
      getToken(email, password).then((token) => {
        const headers = {
          'Authorization': 'Bearer '+ token
        };
        lokka = new Lokka({transport: new HttpTransport(endpoint, {headers})});
        done();
      });
    });

    var newCrossingId;

    it('should add the crossing', async () => {
      const response = await lokka.send(
        newCrossingMutation,
      {
        communityId: communityId,
        longitude: longitude,
        latitude: latitude
      });

      newCrossingId = response.newCrossing.crossing.id;
      expect(response).not.toBeNull();
    });

    it('the new crossing should show up in the DB', async () => {
      const response = await lokka.send(`
        query ($id: Int!) {
          crossingById(id: $id) {
            name
            humanAddress
            description
            coordinates
            humanCoordinates
            communities {
              nodes {
                id
              }
            }
          }
        }
      `,
      {
        id: newCrossingId
      });

      expect(response).toMatchSnapshot();
    });

  });
}

function shouldFail(email, password, communityId, longitude, latitude, extra_description) {
  describe('as ' + email + ' ' + (extra_description || ''), () => {
    var lokka;

    beforeAll(async (done) => {
      getToken(email, password).then((token) => {
        const headers = {
          'Authorization': 'Bearer '+ token
        };
        lokka = new Lokka({transport: new HttpTransport(endpoint, {headers})});
        done();
      });
    });

    it('should fail to add the crossing', async () => {
      try {
        const response = await lokka.send(
          newCrossingMutation,
        {
          communityId: communityId,
          longitude: longitude,
          latitude: latitude
        });
      } catch(e) {
        expect(e).not.toBeNull();
      }
    });
  });
}

describe('When adding a new crossing', () => {
  shouldWork(superAdminEmail, everyPassword, 1, longitude, latitude);
  shouldWork(superAdminEmail, everyPassword, 2, longitude, latitude);
  shouldWork(communityAdminEmail, everyPassword, 1, longitude, latitude);
  shouldFail(communityAdminEmail, everyPassword, 2, longitude, latitude, "to a different community");
  shouldWork(communityEditorEmail, everyPassword, 1, longitude, latitude);
  shouldFail(communityEditorEmail, everyPassword, 2, longitude, latitude, "to a different community");
  shouldFail(superAdminEmail, everyPassword, 2, null, latitude, "without longitude");
  shouldFail(superAdminEmail, everyPassword, 2, longitude, null, "without latitude");
  shouldFail(superAdminEmail, everyPassword, 2,  null, null, "without either coordinates");
});
