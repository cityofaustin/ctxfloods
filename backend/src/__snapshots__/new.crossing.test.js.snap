// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When adding a new crossing as admin@community.floods  the new crossing should show up in the DB 1`] = `
Object {
  "crossingById": Object {
    "communityCrossingsByCrossingId": Object {
      "nodes": Array [
        Object {
          "communityId": 1,
        },
      ],
    },
    "coordinates": "010100000009FA0B3D627058C03BFC3559A34E3E40",
    "description": "TEST LAND IS MAGIC!",
    "humanAddress": "In test land",
    "humanCoordinates": "30°18'25.848\\"N 97°45'21.586\\"W",
    "name": "New Crossing",
  },
}
`;

exports[`When adding a new crossing as admin@community.floods to a different community should fail to add the crossing 1`] = `[Error: GraphQL Error: Users can only add crossings to their communities]`;

exports[`When adding a new crossing as editor@community.floods  the new crossing should show up in the DB 1`] = `
Object {
  "crossingById": Object {
    "communityCrossingsByCrossingId": Object {
      "nodes": Array [
        Object {
          "communityId": 1,
        },
      ],
    },
    "coordinates": "010100000009FA0B3D627058C03BFC3559A34E3E40",
    "description": "TEST LAND IS MAGIC!",
    "humanAddress": "In test land",
    "humanCoordinates": "30°18'25.848\\"N 97°45'21.586\\"W",
    "name": "New Crossing",
  },
}
`;

exports[`When adding a new crossing as editor@community.floods to a different community should fail to add the crossing 1`] = `[Error: GraphQL Error: Users can only add crossings to their communities]`;

exports[`When adding a new crossing as superadmin@flo.ods  the new crossing should show up in the DB 1`] = `
Object {
  "crossingById": Object {
    "communityCrossingsByCrossingId": Object {
      "nodes": Array [
        Object {
          "communityId": 1,
        },
      ],
    },
    "coordinates": "010100000009FA0B3D627058C03BFC3559A34E3E40",
    "description": "TEST LAND IS MAGIC!",
    "humanAddress": "In test land",
    "humanCoordinates": "30°18'25.848\\"N 97°45'21.586\\"W",
    "name": "New Crossing",
  },
}
`;

exports[`When adding a new crossing as superadmin@flo.ods  the new crossing should show up in the DB 2`] = `
Object {
  "crossingById": Object {
    "communityCrossingsByCrossingId": Object {
      "nodes": Array [
        Object {
          "communityId": 2,
        },
      ],
    },
    "coordinates": "010100000009FA0B3D627058C03BFC3559A34E3E40",
    "description": "TEST LAND IS MAGIC!",
    "humanAddress": "In test land",
    "humanCoordinates": "30°18'25.848\\"N 97°45'21.586\\"W",
    "name": "New Crossing",
  },
}
`;

exports[`When adding a new crossing as superadmin@flo.ods without either coordinates should fail to add the crossing 1`] = `[Error: Invalid status code: 500]`;

exports[`When adding a new crossing as superadmin@flo.ods without latitude should fail to add the crossing 1`] = `[Error: Invalid status code: 500]`;

exports[`When adding a new crossing as superadmin@flo.ods without longitude should fail to add the crossing 1`] = `[Error: Invalid status code: 500]`;
